iothub/service/src/Messaging/MessageFeedbackProcessorClient.cs

```diff
@@ -0,0 +1,156 @@
+    {
+        private readonly string _hostName;
+        private readonly IotHubConnectionProperties _credentialProvider;
+        private readonly IotHubConnection _connection;
+
+        /// <summary>
+        /// The callback to be executed each time message feedback is received from the service.
+        /// </summary>
+        /// <remarks>
+        /// May not be null.
+        /// </remarks>
+        public Func<FeedbackBatch, DeliveryAcknowledgement> _messageFeedbackProcessor;
+
+        /// <summary>
+        /// The callback to be executed when the connection is lost.
+        /// </summary>
+        /// <remarks>
+        /// May not be null.
+        /// </remarks>
+        public Action<Exception> _errorProcessor;
+
+        /// <summary>
+        /// Creates an instance of this class. Provided for unit testing purposes only.
+        /// </summary>
+        protected MessageFeedbackProcessorClient()
+        {
+        }
+
+        internal MessageFeedbackProcessorClient(
+            string hostName,
+            IotHubConnectionProperties credentialProvider,
+            IotHubServiceClientOptions options)
+        {
+            _hostName = hostName;
+            _credentialProvider = credentialProvider;
+            _connection = new IotHubConnection(credentialProvider, options.UseWebSocketOnly, options.TransportSettings, options);
+            FeedbackReceiver = new AmqpFeedbackReceiver(_connection);
+        }
+
+        /// <summary>
+        /// Gets the AmqpFeedbackReceiver which can deliver acknowledgments for messages sent to a device/module from IoT hub.
+        /// </summary>
+        internal AmqpFeedbackReceiver FeedbackReceiver { get; }
+
+        /// <summary>
+        /// Open the FeedbackReceiver instance.
+        /// </summary>
+        public virtual async Task OpenAsync()
+        {
+            if (Logging.IsEnabled)
+                Logging.Enter(this, $"Opening FeedbackReceiver", nameof(OpenAsync));
+            try
+            {
+                await FeedbackReceiver.OpenAsync().ConfigureAwait(false);
+            }
+            catch(Exception ex)
+            {
+                if (Logging.IsEnabled)
+                    Logging.Error(this, $"{nameof(OpenAsync)} threw an exception: {ex}", nameof(OpenAsync));
+                throw;
+            }
+            finally
+            {
+                if (Logging.IsEnabled)
+                    Logging.Exit(this, $"Opening FeedbackReceiver", nameof(OpenAsync));
+            }
+        }
+
+        /// <summary>
+        /// Receive cloud-to-device message feedback.
+        /// </summary>
+        /// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
+        /// <exception cref="OperationCanceledException">If the provided <paramref name="cancellationToken"/> has requested cancellation.</exception>
+        public async Task ReceiveAsync(CancellationToken cancellationToken = default)
```

Comment:
It looks like you are mixing design patterns a bit here. A class like this could go one of 2 ways.

The first way is that the user gives us a callback to execute each time a feedback message is received on the opened connection.

```csharp
public class SampleCode
{
    public async Task main()
    {
        serviceClient.MessageFeedbackProcessorClient.MessageFeedbackProcessor = 
            HandleMessageFeedback;
        await serviceClient.MessageFeedbackProcessorClient.OpenAsync();
    }
    
    public AcknowledgementType HandleMessageFeedback(FeedbackBatch feedback)
    {
        // customer logic for handling the feedback
        return AcknowledgementType.Complete; 
    }
}
```


The second way is to ask the user to constantly poll for feedback messages.

```csharp
public class SampleCode
{
    public async Task main()
    {
        await serviceClient.MessageFeedbackProcessorClient.OpenAsync();

        while (true)
        {
            FeedbackBatch feedbackBatch = await serviceClient.MessageFeedbackProcessorClient.ReceiveAsync();
            if (feedbackBatch != null)
            {
                // do something with the received feedbackBatch
            }
        }
    }
}
```

Reply:
I generally prefer the first pattern, but this could be a design discussion you bring up to the broader team if you feel it is worth talking about in depth. If you do go with the first pattern, then there shouldn't be a ```ReceiveAsync()``` call in your class here

Reply:
@timtay-microsoft  where would we invoke the callback in the first case then?

Reply:
Internally. The customer wouldn't need to do anything other than open the connection

Reply:
While we've been reticent to maintain e2e tests of these two AMQP receiver classes due to how flakey the feature feels, it may be worthwhile to write some just to test out your code locally and get a feel for how usable your APIs are here.

Reply:
Azure AMQP does allow you to assign a callback, which would get triggered any time a message is received on the receiving link. This is the API you'd want to use to be able to set a listener: https://github.com/Azure/azure-amqp/blob/43e7b821e4fe119a4a47fa2240a6a1f8871d0848/src/ReceivingAmqpLink.cs#L56-L66

Our device client cloud-to-device callback API makes use of this feature: https://github.com/Azure/azure-iot-sdk-csharp/blob/05a98010ef1daab9dc3208d3aaf97628e2828dc6/iothub/device/src/Transport/Amqp/AmqpUnit.cs#L302

Reply:
This sounds like the ReceiveAsync vs SetC2dMessageHandlerAsync design choice. Right?

I'd prefer the callback too.

Reply:
> This sounds like the ReceiveAsync vs SetC2dMessageHandlerAsync design choice. Right?

Yes, it's the same discussion we had for receiving c2d messages on the device side

Reply:
@timtay-microsoft added the suggested changes.